polaris-core/src/main/java/org/apache/polaris/core/persistence/cache/IndexedCache.java

```diff
@@ -0,0 +1,232 @@
+
+  private IndexedCache(
+      Caffeine<Object, Object> cacheBuilder, LinkedHashSet<Function<V, K>> indexers) {
+    this.indexes = new ConcurrentHashMap<>();
+    this.locks = Striped.lock(1_024);
+    this.indexers = indexers;
+    this.store =
+        cacheBuilder
+            .evictionListener((key, value, cause) -> indexes.keySet().removeAll(indexes.get(key)))
+            .build();
+  }
+
+  /** Returns the value associated with the key or {@code null} if not found. */
+  V getIfPresent(K key) {
+    var index = indexes.get(key);
+    return (index == null) ? null : store.getIfPresent(index.iterator().next());
+  }
+
+  /**
+   * Returns the value associated with the key, obtaining that value from the {@code loader} if
+   * necessary. The entire method invocation is performed atomically, so the function is applied at
+   * most once per key. As the value may be looked up by alternative keys, those function
+   * invocations may be executed in parallel and will replace any existing mappings when completed.
+   */
+  V get(K key, Supplier<V> loader) {
+    var value = getIfPresent(key);
+    if (value != null) {
+      return value;
+    }
+
+    var lock = locks.get(key);
+    lock.lock();
+    try {
+      value = getIfPresent(key);
+      if (value != null) {
+        return value;
+      }
+
+      value = loader.get();
+      if (value == null) {
+        return null;
+      }
+
+      put(value);
+      return value;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /** Associates the {@code value} with its keys, replacing the old value and keys if present. */
+  private void put(V value) {
+    putIfNewer(value, (oldValue, newValue) -> 1);
+  }
+
+  /**
+   * Associates the {@code newValue} with its keys if it is newer than the existing value according
+   * to the provided {@code comparator}. If the new value is added, the old value and its associated
+   * keys are replaced.
+   *
+   * @param newValue the new value to be added to the cache
+   * @param comparator a comparator to determine if the new value is newer than the existing value.
+   *     The comparator must support `null` values in case no value exist in the cache prior to
+   *     invocation.
+   */
+  private void putIfNewer(V newValue, Comparator<V> comparator) {
+    requireNonNull(newValue);
+    var index = buildIndex(newValue);
+    store
+        .asMap()
+        .compute(
+            index.iterator().next(),
+            (key, oldValue) -> {
+              if (comparator.compare(oldValue, newValue) > 0) {
+                if (oldValue != null) {
+                  indexes
+                      .keySet()
+                      .removeAll(Sets.difference(indexes.get(index.iterator().next()), index));
+                }
+                for (var indexKey : index) {
+                  indexes.put(indexKey, index);
+                }
+                return newValue;
+              } else {
+                return oldValue;
```

Comment:
This branch does not seem possible... The comparator always returns `1`, does it not?

Reply:
My bad, another leftover from the attempt at moving the entity version comparison logic to the IndexedCache...